git_xet/src/lfs_agent_protocol/progress_updater.rs

```diff
@@ -0,0 +1,284 @@
+use std::io::Write;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+
+use super::errors::Result;
+use super::{LFSProtocolResponseEvent, ProgressResponse, to_line_delimited_json_string};
+
+/// A progress updater that send custom transfer agent progress messages to git-lfs.
+///
+/// This updater ensures that these messages contain monotonic increasing progress
+/// regardless of reordered/unordered underlying progress registrations due to scheduling.
+///
+/// The implemention handles concurrent updates correctly and is "wait-free": concurrent callers
+/// to `update_bytes_so_far` don't wait for turns to use the update message channel, only the
+/// first that acquires the lock gets to send a message with the latest progress.
+pub struct ProgressUpdater<W: Write + Send + Sync + 'static> {
+    update_channel: Arc<Mutex<W>>,
+    request_oid: String,
+
+    bytes_so_far: AtomicU64,
+    bytes_last_sent: AtomicU64,
+}
+
+impl<W: Write + Send + Sync + 'static> ProgressUpdater<W> {
+    pub fn new(update_channel: Arc<Mutex<W>>, request_oid: &str) -> Self {
+        Self {
+            update_channel,
+            request_oid: request_oid.to_owned(),
+            bytes_so_far: 0.into(),
+            bytes_last_sent: 0.into(),
+        }
+    }
+
+    pub fn update_bytes_so_far(&self, number: u64) -> Result<()> {
+        // load and check instead of swap to ensure monotonic updates
+        let mut current = self.bytes_so_far.load(Ordering::Relaxed);
+
+        loop {
+            // this update is already late, skip this message
+            if current >= number {
+                return Ok(());
+            }
+
+            match self
+                .bytes_so_far
+                .compare_exchange(current, number, Ordering::Relaxed, Ordering::Relaxed)
+            {
+                // no new update was received, record this
+                Ok(_) => {
+                    break;
+                },
+                // another update was received and recorded, check again
+                Err(v) => {
+                    current = v;
+                },
+            }
+        }
+
+        // now this is a valid update, try send only if channel not busy
+        self.try_send_update_message()?;
```

Comment:
With fetch_max, only need to call this if the value was updated.   Then you don't need to store the previously sent value.